### PR TITLE
[WIP] seperate ST module and add joint-training option

### PR DESCRIPTION
**Pull request by LittleChenCc**
### PR types
 New features 
### PR changes
Models 
### Describe
Separate out the ST module and add the function of ASR joint training (multi-task learning). The model is still training. After that, a simultaneous decoding mode can be applied. I make this PR mainly to obtain the opinions from the branch manager. This module needs to be further optimized and organized. So please do NOT merge it into the remote branch.

### CODE REVIEW

File: deepspeech/exps/u2_st/model.py [hunk 1]
```diff
@@ -0,0 +1,781 @@
...
+        return default
+
+    def __init__(self, config, args):
+        super().__init__(config, args)
+
+    def train_batch(self, batch_index, batch_data, msg):
+        train_conf = self.config.training
+        start = time.time()
+        utt, audio, audio_len, text, text_len = batch_data
+        if isinstance(text, list) and isinstance(text_len, list):
```
Comment (zh794390558) [inline deepspeech/exps/u2_st/model.py]: 加注释说明什么情况会走此分支。

File: deepspeech/exps/u2_st/model.py [hunk 2]
```diff
@@ -0,0 +1,781 @@
...
+                    audio, audio_len, text, text_len, text_transcript,
+                    text_transcript_len)
+            else:
+                loss, st_loss, attention_loss, ctc_loss = self.model(
+                    audio, audio_len, text, text_len)
+            if paddle.isfinite(loss):
+                num_utts = batch[1].shape[0]
+                num_seen_utts += num_utts
+                total_loss += float(st_loss) * num_utts
+                valid_losses['val_loss'].append(float(st_loss))
```
Comment (zh794390558) [inline deepspeech/exps/u2_st/model.py]: why not use `loss`?

File: deepspeech/exps/u2_st/model.py [hunk 3]
```diff
@@ -0,0 +1,781 @@
...
+        return dict(
+            errors_sum=errors_sum,
+            len_refs=len_refs,
+            num_ins=num_ins,  # num examples
+            error_rate=errors_sum / len_refs,
+            error_rate_type=cfg.error_rate_type,
+            num_frames=audio_len.sum().numpy().item(),
+            decode_time=decode_time)
+
+    def compute_translation_metrics(self,
```
Comment (zh794390558) [inline deepspeech/exps/u2_st/model.py]: 加入utts，能和 compute_metrics合并最好。

File: deepspeech/exps/u2_st/model.py [hunk 4]
```diff
@@ -0,0 +1,781 @@
...
+
+    @mp_tools.rank_zero_only
+    @paddle.no_grad()
+    def test(self):
+        assert self.args.result_file
+        self.model.eval()
+        logger.info(f"Test Total Examples: {len(self.test_loader.dataset)}")
+
+        stride_ms = self.test_loader.dataset.stride_ms
+        if self.config.decoding.error_rate_type == 'bleu':
```
Comment (zh794390558) [inline deepspeech/exps/u2_st/model.py]: 合并compute metric后此部分可以合并分支

File: deepspeech/exps/u2_st/model.py [hunk 5]
```diff
@@ -0,0 +1,781 @@
...
+        # train/valid dataset, return token ids
+        Dataset = TripletManifestDataset if config.model.model_conf.asr_weight > 0. else ManifestDataset
+        config.data.manifest = config.data.train_manifest
+        train_dataset = Dataset.from_config(config)
+
+        config.data.manifest = config.data.dev_manifest
+        dev_dataset = Dataset.from_config(config)
+
+        if config.collator.raw_wav:
+            TestCollator = Collator = SpeechCollator
```
Comment (zh794390558) [inline deepspeech/exps/u2_st/model.py]: 这部分看着有些乱。

File: deepspeech/frontend/utility.py [hunk 6]
```diff
@@ -80,7 +80,7 @@ def read_manifest(
             token_len / feat_len >= min_output_input_ratio,
             token_len / feat_len <= max_output_input_ratio,
         ]
-        if all(conditions):
+        if all(conditions) or keep_all:
```
Comment (zh794390558) [inline deepspeech/frontend/utility.py]: 默认就是keep all，所以可以去除keep_all

File: deepspeech/io/collator_st.py [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+        if 'tar2object' not in self._local_data.__dict__:
+            self._local_data.tar2object = {}
+        if tarpath not in self._local_data.tar2info:
+            object, infoes = self._parse_tar(tarpath)
+            self._local_data.tar2info[tarpath] = infoes
+            self._local_data.tar2object[tarpath] = object
+        return self._local_data.tar2object[tarpath].extractfile(
+            self._local_data.tar2info[tarpath][filename])
+
+    def process_utterance(self, audio_file, translation):
```
Comment (zh794390558) [inline deepspeech/io/collator_st.py]: raw_wav=False的时候加个函数处理即可，应该可以写在collactor.py的类中。

File: deepspeech/io/collator_st.py [hunk 8]
```diff
@@ -0,0 +1,611 @@
...
+
+        :param audio_file: Filepath or file object of audio file.
+        :type audio_file: str | file
+        :param translation: translation text.
+        :type translation: str
+        :return: Tuple of audio feature tensor and data of translation part,
+                 where translation part could be token ids or text.
+        :rtype: tuple of (2darray, list)
+        """
+        if isinstance(audio_file, str) and audio_file.startswith('tar:'):
```
Comment (zh794390558) [inline deepspeech/io/collator_st.py]: raw_wav=False时， 232-240可以调整下。

File: deepspeech/io/collator_st.py [hunk 9]
```diff
@@ -0,0 +1,611 @@
...
+            texts.append(tokens)
+            text_lens.append(tokens.shape[0])
+
+        padded_audios = pad_sequence(
+            audios, padding_value=0.0).astype(np.float32)  #[B, T, D]
+        audio_lens = np.array(audio_lens).astype(np.int64)
+        padded_texts = pad_sequence(
+            texts, padding_value=IGNORE_ID).astype(np.int64)
+        text_lens = np.array(text_lens).astype(np.int64)
+        return utts, padded_audios, audio_lens, padded_texts, text_lens
```
Comment (zh794390558) [inline deepspeech/io/collator_st.py]: 可以把return变成dict试下，如果可以的话，就可以统一pipeline，在模型中判断是否有对应的key来走mtl，和普通的st

File: deepspeech/io/collator_st.py [hunk 10]
```diff
@@ -0,0 +1,611 @@
...
+
+        :param audio_file: Filepath or file object of kaldi processed feature.
+        :type audio_file: str | file
+        :param translation: Translation text.
+        :type translation: str
+        :return: Tuple of audio feature tensor and data of translation part,
+                 where translation part could be token ids or text.
+        :rtype: tuple of (2darray, list)
+        """
+        specgram = kaldiio.load_mat(audio_file)
```
Comment (zh794390558) [inline deepspeech/io/collator_st.py]: 只需要兼容process_utterance，就不用写两个类了。可以分离出load_wav, load_feat两个函数。

File: deepspeech/io/collator_st.py [hunk 11]
```diff
@@ -0,0 +1,611 @@
...
+        padded_audios = pad_sequence(
+            audios, padding_value=0.0).astype(np.float32)  #[B, T, D]
+        audio_lens = np.array(audio_lens).astype(np.int64)
+        padded_translation = pad_sequence(
+            translation_text, padding_value=IGNORE_ID).astype(np.int64)
+        translation_lens = np.array(translation_text_lens).astype(np.int64)
+        padded_transcription = pad_sequence(
+            transcription_text, padding_value=IGNORE_ID).astype(np.int64)
+        transcription_lens = np.array(transcription_text_lens).astype(np.int64)
+        return utts, padded_audios, audio_lens, (
```
Comment (zh794390558) [inline deepspeech/io/collator_st.py]: 可以是下Collator和Dataset都返回dict来处理，判断key是否存在。模型text时asr的，st_text是translation的。

Comment (zh794390558) [reply deepspeech/io/collator_st.py]: 验证通过后把asr的对应模型输入修改即可。

File: deepspeech/io/dataset.py [hunk 12]
```diff
@@ -96,7 +99,9 @@ def __init__(self,
             max_output_len=max_output_len,
             min_output_len=min_output_len,
             max_output_input_ratio=max_output_input_ratio,
-            min_output_input_ratio=min_output_input_ratio)
+            min_output_input_ratio=min_output_input_ratio,
```
Comment (zh794390558) [inline deepspeech/io/dataset.py]: 如果需要dev,test不过滤的话在 trainer里配置max_output_len,...,等参数为默认值即可。不需要keep_all

File: deepspeech/io/dataset.py [hunk 13]
```diff
@@ -105,3 +110,14 @@ def __len__(self):
...
+
+
+class TripletManifestDataset(ManifestDataset):
+    """
+        For Joint Training of Speech Translation and ASR.
+    """
+
+    def __getitem__(self, idx):
+        instance = self._manifest[idx]
+        return instance["utt"], instance["feat"], instance["text"], instance[
```
Comment (zh794390558) [inline deepspeech/io/dataset.py]: 通过判断key的方式合并TripletManifestDataset和 ManifestDataset， manifest的文件可以加入st_text这样的key.

File: examples/mustc/conf/transformer.yaml [hunk 14]
```diff
@@ -0,0 +1,107 @@
+# https://yaml.org/type/float.html
+data:
```
Comment (zh794390558) [inline examples/mustc/conf/transformer.yaml]: 配置得修改了。

File: examples/mustc/conf/transformer.yaml [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+  scheduler: warmuplr     # pytorch v1.1.0+ required
+  scheduler_conf:
+    warmup_steps: 25000
+    lr_decay: 1.0
+  log_interval: 100
+
+
+decoding:
+  batch_size: 64
+  error_rate_type: bleu
```
Comment (zh794390558) [inline examples/mustc/conf/transformer.yaml]: error_rate_type可以修改成metric_type， 默认越小越好，bleu就得取负

Comment (zh794390558) [reply examples/mustc/conf/transformer.yaml]: 修改的话，所有的地方都要修改。

File: examples/mustc/conf/transformer_joint_noam.yaml [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+  max_output_input_ratio: 20.0
+
+collator:
+  vocab_filepath: data/joint.vocab.txt
+  unit_type: 'spm'
+  spm_model_prefix: data/joint.bpe
+  mean_std_filepath: ""
+  # augmentation_config: conf/augmentation.json
+  batch_size: 10
+  raw_wav: False  # use raw_wav or kaldi feature
```
Comment (zh794390558) [inline examples/mustc/conf/transformer_joint_noam.yaml]: 利用此参数

File: examples/mustc/local/train.sh [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+fi
+
+ngpu=$(echo $CUDA_VISIBLE_DEVICES | awk -F "," '{print NF}')
+echo "using $ngpu gpus..."
+
+config_path=$1
+ckpt_name=$2
+
+device=gpu
+if [ ngpu == 0 ];then
```
Comment (zh794390558) [inline examples/mustc/local/train.sh]: ${ngpu} 都要修改下

File: deepspeech/models/u2_st.py [hunk 18]
```diff
@@ -0,0 +1,1078 @@
...
+        # note that eos is the same as sos (equivalent ID)
+        self.sos = vocab_size - 1
+        self.eos = vocab_size - 1
+        self.vocab_size = vocab_size
+        self.ignore_id = ignore_id
+        self.ctc_weight = ctc_weight
+        self.asr_weight = asr_weight
+
+        self.encoder = encoder
+        self.st_decoder = st_decoder
```
Comment (zh794390558) [inline deepspeech/models/u2_st.py]: 参数不共用吗？

File: deepspeech/models/u2_st.py [hunk 19]
```diff
@@ -0,0 +1,1078 @@
...
+
+        if config is not None:
+            config.merge_from_other_cfg(default)
+        return default
+
+    def __init__(self,
+                 vocab_size: int,
+                 encoder: TransformerEncoder,
+                 st_decoder: TransformerDecoder,
+                 decoder: TransformerDecoder=None,
```
Comment (zh794390558) [inline deepspeech/models/u2_st.py]: 可以考虑用decoders = List[]

Comment (zh794390558) [reply deepspeech/models/u2_st.py]: st_loss, asr_loss计算就可以统一了，shared与否根据传入的deocder是否是同一个决定。

Comment (zh794390558) [reply deepspeech/models/u2_st.py]: 默认顺序asr, st这样？

File: deepspeech/models/u2_st.py [hunk 20]
```diff
@@ -0,0 +1,1078 @@
...
+                loss_asr = loss_ctc
+            else:
+                loss_asr = self.ctc_weight * loss_ctc + (1 - self.ctc_weight
+                                                         ) * loss_att
+            loss = self.asr_weight * loss_asr + (1 - self.asr_weight) * loss_st
+        else:
+            loss = loss_st
+        return loss, loss_st, loss_att, loss_ctc
+
+    def _calc_st_loss(
```
Comment (zh794390558) [inline deepspeech/models/u2_st.py]: _calc_st_loss 和 _calc_att_loss 应该是一个loss

File: deepspeech/models/u2_st.py [hunk 21]
```diff
@@ -0,0 +1,1078 @@
...
+from deepspeech.utils.tensor_utils import pad_sequence
+from deepspeech.utils.tensor_utils import th_accuracy
+from deepspeech.utils.utility import log_add
+
+__all__ = ["U2STModel", "U2STInferModel"]
+
+logger = Log(__name__).getlog()
+
+
+class U2STBaseModel(nn.Module):
```
Comment (zh794390558) [inline deepspeech/models/u2_st.py]: 考虑集成 asr的 BaseModel，省去共用代码，需要自定义的可以覆盖原有接口。

File: deepspeech/exps/u2_st/model.py (hunk 2 above)
Comment (LittleChenCc) [reply deepspeech/exps/u2_st/model.py]: Since the actual goal is to record the best checkpoints for translation. I'll consider using BLEU as a record indicator directly.

File: deepspeech/exps/u2_st/model.py (hunk 5 above)
Comment (LittleChenCc) [reply deepspeech/exps/u2_st/model.py]: 这个是有点乱。我还需要把非kaldi的mtl loader补上。

File: deepspeech/exps/u2_st/model.py (hunk 2 above)
Comment (zh794390558) [reply deepspeech/exps/u2_st/model.py]: 默认用val_loss，选择最好。

File: deepspeech/models/u2_st.py (hunk 20 above)
Comment (zh794390558) [reply deepspeech/models/u2_st.py]: u2_st.py 可以考虑支持 multi decoder，包括atten和ctc的
